### PR TITLE
Add initial codegen documentation

### PR DESCRIPTION
**Pull request by igneel64**

### CODE REVIEW

File: docs/collecting-data/code-generation/snowtype-config/index.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+sidebar_position: 3
+---
+
+The Snowtype CLI configuration will be saved in a`snowtype.config.json` file after initialization. **We highly recommend you keep this file in the root of your project folder.**
+
+## Attributes in `snowtype.config.json`
+
+### `igluCentralSchemas`
+
+The schema tracking URLs from schemas available in [Iglu Central](http://iglucentral.com/).
```
Comment (mscwilson) [inline docs/collecting-data/code-generation/snowtype-config/index.md:12]:
```suggestion
The URLs for schemas available in [Iglu Central](http://iglucentral.com/).
```

File: docs/collecting-data/code-generation/snowtype-config/index.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+### `dataStructures`
+
+The schema tracking URLs from Data Structures published in the Console.
```
Comment (mscwilson) [inline docs/collecting-data/code-generation/snowtype-config/index.md:16]:
```suggestion
The URLs for Data Structures published in the Console.
```

File: docs/collecting-data/code-generation/snowtype-config/index.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+### `trackingScenarioIds`
+
+The Tracking Scenario IDs you wish to generate tracking code for. The Tracking Scenario ID is a UUID that can be retrieved as the final part of the URL when visiting a tracking scenario main page.
+
+### `organizationId`
+
+The Organization ID for your Snowplow BDP account. The Organization ID is a UUID that can be retrieved from the URL immediately following the .com when visiting console.
```
Comment (mscwilson) [inline docs/collecting-data/code-generation/snowtype-config/index.md:24]: link to Console?

File: docs/collecting-data/code-generation/using-the-cli/index.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+sidebar_position: 2
+---
+
+## Working with the Snowtype CLI
+
+The Snowtype CLI is a tool which aims to speed up tracking implementations, provide type safety and inline documentation for developers and ultimately reduce the number of erroneous events. By integrating this tool in the development workflow we introduce a way to connect the additions and updates done in a Snowplow implementation with the corresponding tracking code of the project.
+
+## Authenticating with the Console
+
+It is required for the Snowtype CLI to work properly, that you have a Console API key available. You can find your own or create one in the Console [credentials page](https://console.snowplowanalytics.com/credentials).
```
Comment (mscwilson) [inline docs/collecting-data/code-generation/using-the-cli/index.md:12]:
```suggestion
A Console API key is required for the Snowtype CLI to work properly. You can find your own or create one in the Console [credentials page](https://console.snowplowanalytics.com/credentials).
```

File: docs/collecting-data/code-generation/using-the-cli/index.md [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+From there, the ways for the CLI to recognize the key are either through the global `-k, --apiKey` option or the `SNOWPLOW_CONSOLE_API_KEY` environment variable. The Snowtype CLI does automatically read from a `.env` file as well.
```
Comment (mscwilson) [inline docs/collecting-data/code-generation/using-the-cli/index.md:14]: where does the `.env` file need to be for this to work?

File: docs/collecting-data/code-generation/using-the-cli/index.md [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+**Recommended:** We recommend that you use the `SNOWPLOW_CONSOLE_API_KEY` environment variable.
+
+## Generate tracking code
+
+The CLI will generate tracking code using a valid Snowtype configuration file with the `snowtype generate command`. The code that will be generated, depending on the language, will have all the required types used in schemas and Tracking Scenarios together with methods/classes that allows for tracking these.
+
+The code generated by the CLI is not minified and contains inline documentation for methods, classes and types. If needed you can modify it in any way it suits your project.
```
Comment (mscwilson) [inline docs/collecting-data/code-generation/using-the-cli/index.md:22]: note that any changes will be overwritten next time the generator is run? Should also note somewhere that the classes etc will likely come out in a different order every time it's run

File: docs/collecting-data/code-generation/using-the-cli/index.md [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+### Contents
+The contents of a generated file from the Snowtype CLI will be:
+- Types/Interfaces/Classes for each schema that relates to the Data Structures, Iglu Central Schemas and Tracking Scenarios selected.
+- For each schema:
```
Comment (mscwilson) [inline docs/collecting-data/code-generation/using-the-cli/index.md:27]: links to SelfDescribing JSON/context entity and Selfdescribing event explanations needed here. This section is key so could do with some expansion. Example output from a schema to the codegen output for each language would be great

File: docs/collecting-data/code-generation/using-the-cli/index.md [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    - A method/class to instantiate the structure as a Self Describing JSON. _This is particularly useful to add entities as extra context on events._
+    - A method that sends a Self Describing Event with the schema as the main event entity.
```
Comment (mscwilson) [inline docs/collecting-data/code-generation/using-the-cli/index.md:29]: what does this mean?

File: docs/collecting-data/code-generation/using-the-cli/index.md [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+Following is the set of available trackers and languages the Snowtype CLI currently can work with. This list is also the source of truth for valid keys in the `tracker` and `language` attributes of the Snowtype configuration file.
+
+| **Tracker** | **Language/s** |
+| --- | --- |
+| `@snowplow/browser-tracker` | javascript, typescript |
+| `@snowplow/node-tracker` | javascript, typescript |
+| `@snowplow/javascript-tracker` | javascript |
+| `snowplow_golang_tracker` | go |
+| `snowplow_ios_tracker` | swift |
+| `snowplow_android_tracker` | kotlin |
```
Comment (mscwilson) [inline docs/collecting-data/code-generation/using-the-cli/index.md:47]:
```suggestion
| `snowplow-android-tracker` | kotlin |
```

File: docs/collecting-data/code-generation/using-the-cli/index.md [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+Following is the set of available trackers and languages the Snowtype CLI currently can work with. This list is also the source of truth for valid keys in the `tracker` and `language` attributes of the Snowtype configuration file.
+
+| **Tracker** | **Language/s** |
+| --- | --- |
+| `@snowplow/browser-tracker` | javascript, typescript |
+| `@snowplow/node-tracker` | javascript, typescript |
+| `@snowplow/javascript-tracker` | javascript |
+| `snowplow_golang_tracker` | go |
+| `snowplow_ios_tracker` | swift |
```
Comment (mscwilson) [inline docs/collecting-data/code-generation/using-the-cli/index.md:46]:
```suggestion
| `snowplow-ios-tracker` | swift |
```

**Review by mscwilson — COMMENTED** [review]
Good start. Could do with more examples

File: docs/collecting-data/code-generation/snowtype-config/index.md (hunk 1 above)
Comment (igneel64) [reply docs/collecting-data/code-generation/snowtype-config/index.md:12]: Would keep the `schema tracking URLs` since it is the title of the value we need.

File: docs/collecting-data/code-generation/snowtype-config/index.md (hunk 2 above)
Comment (igneel64) [reply docs/collecting-data/code-generation/snowtype-config/index.md:16]: ^^

File: docs/collecting-data/code-generation/snowtype-config/index.md (hunk 1 above)
Comment (mscwilson) [reply docs/collecting-data/code-generation/snowtype-config/index.md:12]: where's that title used? I haven't seen it before. It feels hard to parse. Like, what is a `tracking URL`? Does this sentence mean "the schema that tracks URLs from schemas in Iglu Central"??
I would just call them `schema URIs`

Comment (igneel64) [reply docs/collecting-data/code-generation/snowtype-config/index.md:12]: image

Comment (mscwilson) [reply docs/collecting-data/code-generation/snowtype-config/index.md:12]: Oh, in that case, let's add a labelled screenshot like this to the page

Comment (mscwilson) [reply docs/collecting-data/code-generation/snowtype-config/index.md:12]: Or put `schema tracking URLs` as inline code

File: docs/collecting-data/code-generation/using-the-cli/index.md (hunk 7 above)
Comment (igneel64) [reply docs/collecting-data/code-generation/using-the-cli/index.md:27]: > Example output from a schema to the codegen output for each language would be great

This will be added moving forwards.

File: docs/collecting-data/code-generation/using-the-cli/index.md (hunk 8 above)
Comment (igneel64) [reply docs/collecting-data/code-generation/using-the-cli/index.md:29]: trackX, would send a self describing event with the X schema as the event data information.
We can consider X as the entity describing the event and the rest entities as context of the event.

Comment (mscwilson) [reply docs/collecting-data/code-generation/using-the-cli/index.md:29]: let's not add more complexity to the `context`/`entity` naming problem 🙈 We should only use "entity" to mean "context entity"

File: docs/collecting-data/code-generation/commands/index.md [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+ -  `-i, --organizationId` Organization ID.
+ -  `-t, --tracker` Tracker to use. [See available](../using-the-cli/index.md#available-trackerslanguages)
+ -  `-l, --language` Language to use. [See available](../using-the-cli/index.md#available-trackerslanguages)
+ -  `-o, --outpath` Output path.
+
+### `snowtype generate`
+
+Generates the code based on configuration on the `snowtype.config.json` file. Can generate/modify the `.snowtype-lock.json` file.
+
+### `snowtype update`
```
[the remark this answers is not in the document]
Comment (igneel64) [reply docs/collecting-data/code-generation/commands/index.md:31]: This page is more of a command reference, no essence of user flow _(which is still being worked on)_. Since commands are, and probably be, just a few, I don't feel there is particular reason to order them based on something. Maybe just `init` and `generate` which would be the most interesting

**Review by adatzer — APPROVED** [review]
LGTM!!
